---
title: "Combined Samples - 06 coppaFISH Genes"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "results", "comb", "comb_06coppafish.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-snRNAseq](https://github.com/eturkes/tau-mutant-snRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024-2025  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(edgeR::cpm, DT::JS, .quiet = TRUE)
packages <- c("Seurat", "scuttle", "edgeR", "DT", "tidyr", "plotly", "reshape2")
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 96)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
# -----------------

# Useful variables.
# -----------------
step <- 2

cache_dir <- file.path("..", "..", "cache", "comb", paste0("step", "0", step))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------

# Read data.
# ----------
seurat_orig <- readRDS(file.path(cache_dir, "annotated_seurat.rds"))
seurat_orig <- seurat_orig[
  , seurat_orig$genotype == "MAPTKI" | seurat_orig$genotype == "S305N"
]
seurat_orig$genotype <- factor(seurat_orig$genotype)

gene_anno <- data.frame(
  symbol = unlist(seurat_orig@misc), ensembl = rownames(seurat_orig[["RNA"]])
)
# ----------
```

# MAPTKI vs. S305N All Cell-types

```{r}
seurat <- seurat_orig

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N All Cell-types", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N All Cell-types", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N Glutamatergic Neurons

```{r}
seurat <- seurat_orig[
  , grep("Glutamatergic", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Glutamatergic Neurons", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Glutamatergic Neurons", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N GABAergic Neurons

```{r}
seurat <- seurat_orig[
  , grep("GABAergic", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N GABAergic Neurons", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N GABAergic Neurons", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N Microglia

```{r}
seurat <- seurat_orig[
  , grep("Microglia", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Microglia", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Microglia", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N Astrocytes

```{r}
seurat <- seurat_orig[
  , grep("Astrocyte", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Astrocytes", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Astrocytes", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N Vascular Cells

```{r}
seurat <- seurat_orig[
  , grep("Vascular", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Vascular Cells", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Vascular Cells", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N Oligodendrocytes

```{r}
seurat <- seurat_orig[
  , grep("Oligodendrocyte", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Oligodendrocytes", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N Oligodendrocytes", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# MAPTKI vs. S305N OPCs

```{r}
seurat <- seurat_orig[
  , grep("OPC", x = seurat_orig$level1_subclusters)
]
seurat$level1_subclusters <- factor(seurat$level1_subclusters)

deg <- suppressWarnings(as.SingleCellExperiment(seurat, assay = "RNA"))
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)

dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$genotype,
  min.count = 10, min.total.count = 15
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(dge, log = TRUE)

design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$X <- gene_anno_sub$symbol

colnames(results)[1] <- "Gene"
colnames(results)[3] <- "Log2FC_positive_is_up_in_S305N_vs_MAPTKI"

deg_counts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "counts", statistics = "mean"
  )
)
deg_logcounts <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype"],
    use_exprs_values = "logcounts", statistics = "mean"
  )
)

highlight <- c("Cryab", "Plp1")
cat("Genes of Interest:", highlight)
```

## Pseudobulked log2 CPM Gene Expression

```{r}
df <- as.data.frame(logcounts(deg_logcounts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Normalised Expression in MAPTKI") +
  ylab("Normalised Expression in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N OPCs", nrow(deg),
      "Genes: Pseudobulked log2 CPM Expression", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Pseudobulked Raw Gene Counts

```{r}
df <- as.data.frame(counts(deg_counts))
rownames(df) <- results$Gene
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Gene of Interest` <- ifelse(df$Gene %in% highlight, yes = "Yes", "No")

plot <- ggplot(df, aes(MAPTKI, S305N, label = Gene)) +
  geom_point(aes(color = `Gene of Interest`)) +
    geom_abline(intercept = 0, slope = 1) +
  xlab("Raw Counts in MAPTKI") +
  ylab("Raw Counts in S305N") +
  ggtitle(
    paste(
      "MAPTKI vs. S305N OPCs", nrow(deg),
      "Genes: Pseudobulked Raw Counts", sep = " "
    )
  ) +
  theme_light()
ggplotly(plot)
```

## Differential Gene Expression Stats

```{r}
suppressWarnings(datatable_download_exp(results))
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
