---
title: "Tau Mutant snRNAseq Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "GSEABase", "SingleCellExperiment", "scales", "ComplexHeatmap", "circlize",
  "limma", "DT"
)
lapply(packages, FUN = library, character.only = TRUE)

datatable_download_exp <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}

`%notin%` <- Negate(`%in%`)

data_dir <- file.path("..", "..", "cache", "comb", "step03")

gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
deg_data <- readRDS(file.path(data_dir, "deg_data.rds"))
deg_list <- readRDS(file.path(data_dir, "deg_list.rds"))
gse_list <- readRDS(file.path(data_dir, "gse_list.rds"))
corrections <- readRDS(file.path(data_dir, "corrections.rds"))

ctypes <- c(
  "Astrocytes", "Oligodendrocytes", "Microglia", "Glutamatergic CA3",
  "Glutamatergic CA1", "Glutamatergic L2/3 ENT", "GABAergic Sst"
)

anno_color <- hue_pal()(4)

top_anno <- HeatmapAnnotation(
  Batch = deg_data[[1]]$batch,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = sub(
      pattern = "_", replacement = " ", x = levels(deg_data[[1]]$genotype)
    ),
    labels_gp = gpar(col = "black", fontsize = 11),
    height = unit(5, "mm")
  ),
  col = list(
    Batch = c(
      "batch01" = anno_color[1], "batch02" = anno_color[2],
      "batch03" = anno_color[3], "batch04" = anno_color[4]
    )
  ),
  simple_anno_size = unit(3, "mm"),
  show_legend = c("Batch" = FALSE),
  annotation_name_gp = gpar(fontsize = 9)
)

deg <- deg_data[[1]]
design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  P301S_MAPTKI =
    deg.genotypeP301S -
    deg.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    deg.genotypeNLGF_MAPTKI -
    deg.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    deg.genotypeNLGF_S305N -
    deg.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    deg.genotypeNLGF_P301S -
    deg.genotypeMAPTKI,
  levels = design
)

ctype_name <- vector("list", length = 7)
deg_corrected <- vector("list", length = 7)
deg_range <- vector("list", length = 7)
results_list <- vector("list", length = 7)
sig <- vector("list", length = 7)
sig_uniq <- vector("list", length = 7)
```

Astrocytes
===

```{r, include = FALSE}
idx <- 1
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

Oligodendrocytes
===

```{r, include = FALSE}
idx <- 2
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

Microglia
===

```{r, include = FALSE}
idx <- 3
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

Glutamatergic CA3
===

```{r, include = FALSE}
idx <- 4
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

Glutamatergic CA1
===

```{r, include = FALSE}
idx <- 5
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

Glutamatergic L2/3 ENT
===

```{r, include = FALSE}
idx <- 6
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

GABAergic Sst
===

```{r, include = FALSE}
idx <- 7
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_list[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) +
  nrow(gse_list[[7]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) +
  nrow(gse_list[[7]]) + nrow(deg_list[[7]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_list[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_list[[idx]]) <- results_list[[idx]]$X
results_list[[idx]] <- results_list[[idx]][ , -1]
results_list[[idx]] <- results_list[[idx]][ , 1:22]
tests <- results_list[[idx]][ , 17:21]
results_list[[idx]] <- results_list[[idx]][
  order(results_list[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_list[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_list[[idx]]), gene_anno_sub$ensembl),
]
results_list[[idx]]$Gene <- gene_anno_sub$symbol
results_list[[idx]] <- results_list[[idx]][
  , c(ncol(results_list[[idx]]), 1:(ncol(results_list[[idx]]) - 1))
]

sig_tmp <- ifelse(results_list[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_list[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction
sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig[[idx]] <- do.call(cbind, sig[[idx]])
rownames(sig[[idx]]) <- results_list[[idx]]$Gene[
  match(rownames(sig[[idx]]), rownames(results_list[[idx]]))
]
sig_uniq[[idx]] <- unique(sig[[idx]])
```

```{r}
selectizeInput(
  paste0("select", idx),
  label = "Select Genes:",
  choices = results_list[[idx]]$Gene,
  selected = rownames(sig_uniq[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig[[idx]][rownames(sig[[idx]]) %in% gene_anno_sub$symbol, ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Results Table

```{r}
datatable_download_exp(results_list[[idx]])
```

```{r, include = FALSE}
rm(deg, deg_data, deg_list, gse_list, corrections)
gc()
```
