---
title: "Tau Mutant snRNAseq Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "GSEABase", "SingleCellExperiment", "scales", "ComplexHeatmap", "circlize",
  "limma", "DT"
)
lapply(packages, FUN = library, character.only = TRUE)

datatable_download_exp_15 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}
datatable_download_exp_5 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 5,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}

`%notin%` <- Negate(`%in%`)

data_dir <- file.path("..", "..", "cache", "comb", "step03")
gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_sets_keep <- readRDS(file.path(data_dir, "gene_sets_keep.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
deg_data <- readRDS(file.path(data_dir, "deg_data.rds"))
gse_data <- readRDS(file.path(data_dir, "gse_data.rds"))
deg_list <- readRDS(file.path(data_dir, "deg_list.rds"))
gse_list <- readRDS(file.path(data_dir, "gse_list.rds"))
corrections <- readRDS(file.path(data_dir, "corrections.rds"))

ctypes <- c(
  "Astrocytes", "Oligodendrocytes", "Microglia", "Glutamatergic CA3",
  "Glutamatergic CA1", "Glutamatergic L2/3 ENT", "GABAergic Sst"
)

anno_color <- hue_pal()(4)
top_anno <- HeatmapAnnotation(
  Batch = deg_data[[1]]$batch,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = sub(
      pattern = "_", replacement = " ", x = levels(deg_data[[1]]$genotype)
    ),
    labels_gp = gpar(col = "black", fontsize = 11),
    height = unit(5, "mm")
  ),
  col = list(
    Batch = c(
      "batch01" = anno_color[1], "batch02" = anno_color[2],
      "batch03" = anno_color[3], "batch04" = anno_color[4]
    )
  ),
  simple_anno_size = unit(3, "mm"),
  show_legend = c("Batch" = FALSE),
  annotation_name_gp = gpar(fontsize = 9)
)

deg <- deg_data[[1]]
design_deg <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design_deg) <- make.names(colnames(design_deg))
cont_mat_deg <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  P301S_MAPTKI =
    deg.genotypeP301S -
    deg.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    deg.genotypeNLGF_MAPTKI -
    deg.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    deg.genotypeNLGF_S305N -
    deg.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    deg.genotypeNLGF_P301S -
    deg.genotypeMAPTKI,
  levels = design_deg
)

gse <- gse_data[[1]]
design_gse <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design_gse) <- make.names(colnames(design_gse))
cont_mat_gse <- makeContrasts(
  S305N_MAPTKI =
    gse.genotypeS305N -
    gse.genotypeMAPTKI,
  P301S_MAPTKI =
    gse.genotypeP301S -
    gse.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    gse.genotypeNLGF_MAPTKI -
    gse.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    gse.genotypeNLGF_S305N -
    gse.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    gse.genotypeNLGF_P301S -
    gse.genotypeMAPTKI,
  levels = design_gse
)

ctype_name <- vector("list", length = 7)
deg_corrected <- vector("list", length = 7)
gse_corrected <- vector("list", length = 7)
deg_range <- vector("list", length = 7)
gse_range <- vector("list", length = 7)
results_deg <- vector("list", length = 7)
results_gse <- vector("list", length = 7)
sig_deg <- vector("list", length = 7)
sig_gse <- vector("list", length = 7)
sig_uniq_deg <- vector("list", length = 7)
sig_uniq_gse <- vector("list", length = 7)
```

Astrocytes
===

```{r, include = FALSE}
idx <- 1
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- 1
length2 <- nrow(gse_list[[1]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

Oligodendrocytes
===

```{r, include = FALSE}
idx <- 2
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

Microglia
===

```{r, include = FALSE}
idx <- 3
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

Glutamatergic CA3
===

```{r, include = FALSE}
idx <- 4
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

Glutamatergic CA1
===

```{r, include = FALSE}
idx <- 5
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

Glutamatergic L2/3 ENT
===

```{r, include = FALSE}
idx <- 6
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

GABAergic Sst
===

```{r, include = FALSE}
idx <- 7
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) +
  nrow(gse_list[[7]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) +
  nrow(gse_list[[7]]) + nrow(deg_list[[7]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_deg[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:22]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]]) + nrow(gse_list[[2]]) +
  nrow(deg_list[[2]]) + nrow(gse_list[[3]]) + nrow(deg_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(deg_list[[4]]) + nrow(gse_list[[5]]) +
  nrow(deg_list[[5]]) + nrow(gse_list[[6]]) + nrow(deg_list[[6]]) +
  nrow(gse_list[[7]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results_gse[[idx]][ , col] <- corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:22]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 19:23] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 4:8] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$Gene,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]],
    ]
    sig_tmp <- sig_tmp[
      match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
    ]

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1350,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_5(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol,
    ]
    sig_tmp <- sig_tmp[match(gene_anno_sub$symbol, rownames(sig_tmp)), ]

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(mat))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 850,
  height = 650,
  res = 96
)
```

</div>
</div>

```{r, include = FALSE}
rm(deg, deg_data, deg_list, gse, gse_data, gse_list, corrections)
gc()
```
