---
title: "Tau Mutant snRNAseq"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
packages <- c(
  "GSEABase", "SingleCellExperiment", "scales", "ComplexHeatmap", "DT",
  "circlize", "networkD3", "limma", "purrr", "dplyr"
)
lapply(packages, FUN = library, character.only = TRUE)

`%notin%` <- Negate(`%in%`)

datatable_download <- function(dt) {
  datatable(
    dt,
    list(
      scrollX = TRUE, dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      )
    ),
    extensions = "Buttons"
  )
}

data_dir <- file.path("..", "..", "cache", "comb", "step03")

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_sets_keep <- readRDS(file.path(data_dir, "gene_sets_keep.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
gse_data <- readRDS(file.path(data_dir, "gse_data.rds"))
deg_data <- readRDS(file.path(data_dir, "deg_data.rds"))
gse_list <- readRDS(file.path(data_dir, "gse_list.rds"))
deg_list <- readRDS(file.path(data_dir, "deg_list.rds"))
corrections <- readRDS(file.path(data_dir, "corrections.rds"))

map <- setNames(gene_anno$symbol, gene_anno$ensembl)

ctypes <- list(
  Astrocytes = c(
    "Astrocyte 01", "Astrocyte 02", "Astrocyte 03",
    "Astro Homeostatic 01", "Astro DAA 01"
  ),
  Oligodendrocytes = c(
    "Oligodendrocyte 01", "Oligodendrocyte 02", "Oligodendrocyte 03",
    "Oligodendrocyte 04", "Oligodendrocyte 05",
    "Oligo Homeostatic 01", "Oligo DAO 01"
  ),
  Microglia = c(
    "Microglia 01", "Microglia 02", "Micro Homeostatic 01",
    "Micro DAM 01", "Micro DAM 02"
  ),
  `Glutamatergic CA3` = c("Glut CA3 01", "Glut CA3 02", "Glut CA3 03"),
  `Glutamatergic CA1` = c("Glut CA1 01"),
  `Glutamatergic L2/3 ENT` = c("Glut L2 ENT 01", "Glut L3 ENT 01"),
  `GABAergic Sst` = c("GABA Sst 01")
)

priority <- c(
  "tau-protein", "tau protein", "neurofi", "microtubule", "dynein", "cytoskele",
  "laminin", "inclusion", "aggrep", "death", "necro", "folding", "folded",
  "synap", "axon", "dendrit", "senescen", "apopto", "exocyt", "endocyt",
  "phagocyt", "proteasom", "lysosom", "endoplasmic", "Golgi", "Hsp",
  "heat shock", "ESCRT", "ubiqui", "heparan", "chondroitin", "sulfotransferase",
  "glycosaminoglycan", "transcription", "translation", "amyloid",
  "apolipoprotein", "lipoprotein", "fatty", "cholesterol", "insulin", "pruning",
  "phosphatidylserine", "complement", "T cell", "B cell", "immun", "inflammat",
  "stress", "cortisol", "interleukin", "leukocyte", "myeloid", "NF-kappaB",
  "toll-like", "macrophage", "astrocyte", "oligodendrocyte", "Schwann", "myelin"
)

anno_color <- hue_pal()(4)
```

Astrocytes
===

```{r}
ctype_idx <- 1

gse <- gse_data[[ctype_idx]]
deg <- deg_data[[ctype_idx]]

gse_corrected <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
deg_corrected <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)

gse_range <- c(
  min(gse_corrected),
  (min(gse_corrected) + max(gse_corrected)) / 2,
  max(gse_corrected)
)
deg_range <- c(
  min(deg_corrected),
  (min(deg_corrected) + max(deg_corrected)) / 2,
  max(deg_corrected)
)
```

Row {.tabset}
---

### Search Genes

```{r}
design <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  P301S_MAPTKI =
    deg.genotypeP301S -
    deg.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    deg.genotypeNLGF_MAPTKI -
    deg.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    deg.genotypeNLGF_S305N -
    deg.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    deg.genotypeNLGF_P301S -
    deg.genotypeMAPTKI,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(data_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(data_dir, "results.tsv"))

idx <- 18
length1 <- nrow(gse_list[[1]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(deg_list[[1]])
length <- length1:length2
for (i in seq_along(corrections)) {
  results[ , idx] <- corrections[[i]][length]
  idx <- idx + 1
}

rownames(results) <- results$X
results <- results[ , -1]
results <- results[ , 1:22]
tests <- results[ , 17:21]
results <- results[order(results$F, decreasing = TRUE), ]
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$symbol <- gene_anno_sub$symbol

results_up <- vector("list", length = 13)
results_down <- vector("list", length = 13)
size <- Inf

results_sub <- results[
  results[ , 17] < 0.05 & results[ , 18] >= 0.05 & results[ , 20] >= 0.05,
]
results_sub <- results_sub[order(results_sub[ , 17]), ]
results_sub_up <- results_sub[
  results_sub[ , 2] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[1]] <- results_sub_up[1:size, ]
  else {
    results_up[[1]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 2] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[1]] <- results_sub_down[1:size, ]
  else {
    results_down[[1]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] < 0.05 & results[ , 18] >= 0.05 & results[ , 20] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(17, 20)])), ]
results_sub_up <- results_sub[
  results_sub[ , 2] > 0 & results_sub[ , 5] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[2]] <- results_sub_up[1:size, ]
  else {
    results_up[[2]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 2] < 0 & results_sub[ , 5] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[2]] <- results_sub_down[1:size, ]
  else {
    results_down[[2]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] < 0.05 & results[ , 21] >= 0.05,
]
results_sub <- results_sub[order(results_sub[ , 18]), ]
results_sub_up <- results_sub[
  results_sub[ , 3] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[3]] <- results_sub_up[1:size, ]
  else {
    results_up[[3]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 3] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[3]] <- results_sub_down[1:size, ]
  else {
    results_down[[3]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] < 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(18, 21)])), ]
results_sub_up <- results_sub[
  results_sub[ , 3] > 0 & results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[4]] <- results_sub_up[1:size, ]
  else {
    results_up[[4]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 3] < 0 & results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[4]] <- results_sub_down[1:size, ]
  else {
    results_down[[4]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] < 0.05 & results[ , 18] < 0.05 &
    results[ , 20] >= 0.05 & results[ , 21] >= 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(17, 18)])), ]
results_sub_up <- results_sub[
  results_sub[ , 2] > 0 & results_sub[ , 3] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[5]] <- results_sub_up[1:size, ]
  else {
    results_up[[5]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 2] < 0 & results_sub[ , 3] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[5]] <- results_sub_down[1:size, ]
  else {
    results_down[[5]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] < 0.05 & results[ , 18] < 0.05 &
    results[ , 20] < 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(17, 18, 20, 21)])), ]
results_sub_up <- results_sub[
  results_sub[ , 2] > 0 & results_sub[ , 3] > 0 &
    results_sub[ , 5] > 0 & results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[6]] <- results_sub_up[1:size, ]
  else {
    results_up[[6]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 2] < 0 & results_sub[ , 3] < 0 &
    results_sub[ , 5] < 0 & results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[6]] <- results_sub_down[1:size, ]
  else {
    results_down[[6]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] < 0.05 &
    results[ , 20] >= 0.05 & results[ , 21] >= 0.05,
]
results_sub <- results_sub[order(results_sub[ , 19]), ]
results_sub_up <- results_sub[
  results_sub[ , 4] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[7]] <- results_sub_up[1:size, ]
  else {
    results_up[[7]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 4] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[7]] <- results_sub_down[1:size, ]
  else {
    results_down[[7]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] < 0.05 &
    results[ , 20] >= 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(19, 21)])), ]
results_sub_up <- results_sub[
  results_sub[ , 4] > 0 & results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[8]] <- results_sub_up[1:size, ]
  else {
    results_up[[8]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 4] < 0 & results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[8]] <- results_sub_down[1:size, ]
  else {
    results_down[[8]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] >= 0.05 &
    results[ , 20] < 0.05 & results[ , 21] >= 0.05,
]
results_sub <- results_sub[order(results_sub[ , 20]), ]
results_sub_up <- results_sub[
  results_sub[ , 5] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[9]] <- results_sub_up[1:size, ]
  else {
    results_up[[9]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 5] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[9]] <- results_sub_down[1:size, ]
  else {
    results_down[[9]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] < 0.05 &
    results[ , 20] < 0.05 & results[ , 21] >= 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(19, 20)])), ]
results_sub_up <- results_sub[
  results_sub[ , 4] > 0 & results_sub[ , 5] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[10]] <- results_sub_up[1:size, ]
  else {
    results_up[[10]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 4] < 0 & results_sub[ , 5] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[10]] <- results_sub_down[1:size, ]
  else {
    results_down[[10]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] < 0.05 &
    results[ , 20] < 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(19, 20, 21)])), ]
results_sub_up <- results_sub[
  results_sub[ , 4] > 0 & results_sub[ , 5] > 0 & results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[11]] <- results_sub_up[1:size, ]
  else {
    results_up[[11]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 4] < 0 & results_sub[ , 5] < 0 & results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[11]] <- results_sub_down[1:size, ]
  else {
    results_down[[11]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] >= 0.05 &
    results[ , 20] < 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(rowSums(results_sub[ , c(20, 21)])), ]
results_sub_up <- results_sub[
  results_sub[ , 5] > 0 & results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[12]] <- results_sub_up[1:size, ]
  else {
    results_up[[12]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 5] < 0 & results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[12]] <- results_sub_down[1:size, ]
  else {
    results_down[[12]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

results_sub <- results[
  results[ , 17] >= 0.05 & results[ , 18] >= 0.05 & results[ , 19] >= 0.05 &
    results[ , 20] >= 0.05 & results[ , 21] < 0.05,
]
results_sub <- results_sub[order(results_sub[ , 21]), ]
results_sub_up <- results_sub[
  results_sub[ , 6] > 0,
]
if (nrow(results_sub_up) > 0) {
  if (nrow(results_sub_up) > size)
    results_up[[13]] <- results_sub_up[1:size, ]
  else {
    results_up[[13]] <- results_sub_up[1:nrow(results_sub_up), ]
  }
}
results_sub_down <- results_sub[
  results_sub[ , 6] < 0,
]
if (nrow(results_sub_down) > 0) {
  if (nrow(results_sub_down) > size)
    results_down[[13]] <- results_sub_down[1:size, ]
  else {
    results_down[[13]] <- results_sub_down[1:nrow(results_sub_down), ]
  }
}

order <- append(
  lapply(results_up, FUN = function(x) x[1, , drop = FALSE]),
  lapply(results_down, FUN = function(x) x[1, , drop = FALSE])
)
results_sub <- results[rownames(results) %in% rownames(do.call(rbind, order)), ]
results_sub <- results_sub[
  match(rownames(do.call(rbind, order)), rownames(results_sub)),
]
results <- results[rownames(results) %notin% rownames(results_sub), ]
results <- rbind(results_sub, results)
```

```{r}
top_anno <- HeatmapAnnotation(
  Batch = gse_data[[1]]$batch,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = levels(gse_data[[1]]$genotype),
    labels_gp = gpar(col = "black", fontsize = 9),
    height = unit(5, "mm")
  ),
  col = list(
    Batch = c(
      "batch01" = anno_color[1], "batch02" = anno_color[2],
      "batch03" = anno_color[3], "batch04" = anno_color[4]
    )
  ),
  simple_anno_size = unit(3, "mm"),
  show_legend = c("Batch" = FALSE),
  annotation_name_gp = gpar(fontsize = 9)
)

selectizeInput(
  "select_genes",
  label = "Select Genes:",
  choices = results$symbol, 
  selected = do.call(rbind, order)$symbol,
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    gene_anno_sub <- gene_anno[gene_anno$symbol %in% input$select_genes, ]
    gene_anno_sub <- gene_anno_sub[
      match(input$select_genes, gene_anno_sub$symbol),
    ]
    mat <- deg_corrected[
      rownames(deg_corrected) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol
    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range, colors = c("blue", "white", "red")),
        column_title = names(ctypes)[ctype_idx],
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 600,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    gene_anno_sub <- gene_anno[gene_anno$symbol %in% input$select_genes, ]
    gene_anno_sub <- gene_anno_sub[
      match(input$select_genes, gene_anno_sub$symbol),
    ]
    mat <- deg_corrected[
      rownames(deg_corrected) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), ]
    rownames(mat) <- gene_anno_sub$symbol
    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = names(ctypes)[ctype_idx],
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        column_split = rep(seq(6), each = 4),
        row_names_gp = gpar(fontsize = 12),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 9),
          grid_height = unit(3, "mm")
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 600,
  res = 96
)
```

</div>
</div>
