---
title: "Combined Samples - 03 GeneFunnel"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "results", "comb", "comb_03genefunnel.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-snRNAseq](https://github.com/eturkes/tau-mutant-snRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(edgeR::cpm, .quiet = TRUE)
packages <- c(
  "GSEABase", "Seurat", "scuttle", "edgeR", "DT", "GSVA", "Rcpp", "biomaRt",
  "IHW"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 96)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
sourceCpp(file.path("..", "..", "src", "calculateScores.cpp"))

`%notin%` <- Negate(`%in%`)
# -----------------

# Useful variables.
# -----------------
step <- 3

cache_dir <- file.path("..", "..", "cache", "comb", paste0("step", "0", step))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------
```

# Prep

```{r}
rds <- file.path(cache_dir, "seurat_orig.rds")
if (file.exists(rds)) {
  seurat_orig <- readRDS(rds)
} else {

  file_list <- list.files(file.path(cache_dir, "..", ".."), recursive = TRUE)
  file_list <- file_list[grep("seurat_prep.rds", file_list)]
  files <- vector("list", length(file_list))
  for (i in seq_along(files)) {
    files[[i]] <- readRDS(file.path(cache_dir, "..", "..", file_list[i]))
    files[[i]][["RNA"]] <- NULL
    files[[i]]@meta.data[grep("SCT", colnames(files[[i]]@meta.data))] <- NULL
  }

  gc()
  seurat_orig <- merge(
    files[[1]], c(files[[2]], files[[3]], files[[4]],
    files[[5]], files[[6]], files[[7]], files[[8]],
    files[[9]], files[[10]], files[[11]], files[[12]],
    files[[13]], files[[14]], files[[15]], files[[16]],
    files[[17]], files[[18]], files[[19]], files[[20]],
    files[[21]], files[[22]], files[[23]], files[[24]]
    )
  )
  rm(files)
  gc()

  seurat_orig <- PrepSCTFindMarkers(seurat_orig, verbose = FALSE)
  gc()

  counts_mat <- GetAssayData(seurat_orig)
  counts_mat[is.na(counts_mat)] <- 0
  seurat_orig <- SetAssayData(seurat_orig, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat_orig, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat_orig <- SetAssayData(
    seurat_orig, slot = "counts", new.data = counts_mat
  )

  seurat_orig$nCount_SCT <- colSums(seurat_orig, layer = "counts")
  seurat_orig$nFeature_SCT <- colSums(counts_mat > 0)

  seurat_anno <- readRDS(
    file.path(cache_dir, "..", "step02", "annotated_seurat.rds")
  )
  seurat_orig[["RNA"]] <- CreateAssay5Object(
    GetAssayData(seurat_anno, assay = "RNA")
  )
  seurat_orig@meta.data <- seurat_anno[[]]
  seurat_orig[["pca"]] <- seurat_anno[["pca"]]
  seurat_orig[["umap"]] <- seurat_anno[["umap"]]
  seurat_orig@misc <- seurat_anno@misc
  seurat_orig@active.ident <- seurat_orig$level2_subclusters

  rm(seurat_anno)
  gc()
  saveRDS(seurat_orig, file = rds)
}
seurat_orig

rds <- file.path(cache_dir, "gene_sets.rds")
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {

  gene_sets <- getGmt(
    file.path(
      "..", "..", "data",
      "gprofiler_mmusculus_ENSG_2024_06_15", "mmusculus.GO.comb.ENSG.gmt"
    )
  )
  keep <- filterGeneSets(gene_sets, min.sz = 2)
  gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

  remove <- unique(unlist(geneIds(gene_sets)))[
    unique(unlist(geneIds(gene_sets))) %notin% rownames(seurat_orig)
  ]
  remove <- which(
    sapply(geneIds(gene_sets), FUN = function(x) any(x %in% remove))
  )
  gene_sets <- gene_sets[-remove]

  for (i in seq_along(gene_sets@.Data)) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets, uniqGenes = unique(unlist(geneIds(gene_sets)))
  )
  gene_sets_keep <- rowSums(gene_sets_keep)
  gene_sets_keep <- gene_sets_keep[order(gene_sets_keep, decreasing = TRUE)]
  gene_sets_keep <- gene_sets[match(names(gene_sets_keep), names(gene_sets))]

  dup <- sapply(
    geneIds(gene_sets_keep), FUN = function(x) paste(sort(x), collapse = "-")
  )
  dup <- which(duplicated(dup))
  gene_sets <- gene_sets_keep[-dup]

  saveRDS(gene_sets, file = rds)
}
gene_sets

rds <- file.path(cache_dir, "gene_sets_keep.rds")
if (file.exists(rds)) {
  gene_sets_keep <- readRDS(rds)
} else {

  keep <- filterGeneSets(gene_sets, min.sz = 2, max.sz = 45)
  gene_sets_keep <- gene_sets[names(gene_sets) %in% names(keep)]

  gene_sets_keep <- gene_sets_keep[
    !grepl("regulation", x = names(gene_sets_keep)) |
      grepl("positive regulation", x = names(gene_sets_keep))
  ]

  gene_sets_keep <- gene_sets_keep[
    !grepl("selection", x = names(gene_sets_keep)) |
      grepl("positive selection", x = names(gene_sets_keep))
  ]

  mart <- useEnsembl("ensembl", "mmusculus_gene_ensembl")
  attributes <- c("ensembl_gene_id", "chromosome_name")
  gene_anno <- getBM(
    attributes, filters = "ensembl_gene_id",
    values = unique(unlist(geneIds(gene_sets_keep))), mart = mart
  )
  remove <- gene_anno[
    gene_anno$chromosome_name == "Y" |
      gene_anno$ensembl_gene_id == "ENSMUSG00000086503" |
      gene_anno$ensembl_gene_id == "ENSMUSG00000085715",
  ]$ensembl_gene_id
  remove <- which(
    sapply(geneIds(gene_sets_keep), FUN = function(x) any(x %in% remove))
  )
  gene_sets_keep <- gene_sets_keep[-remove]

  gene_sets_ordered <- computeGeneSetsOverlapMax(
    gene_sets_keep, uniqGenes = unique(unlist(geneIds(gene_sets_keep)))
  )
  gene_sets_ordered <- rowSums(gene_sets_ordered)
  gene_sets_ordered <- gene_sets_ordered[order(gene_sets_ordered)]
  gene_sets_keep <- gene_sets_keep[
    match(names(gene_sets_ordered), names(gene_sets_keep))
  ]

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets_keep, uniqGenes = unique(unlist(geneIds(gene_sets_keep)))
  )
  gene_sets_keep[upper.tri(gene_sets_keep)] <- 0
  diag(gene_sets_keep) <- 0
  gene_sets_keep <- apply(gene_sets_keep, MARGIN = 1, FUN = max)
  gene_sets_keep <- gene_sets_keep[gene_sets_keep < 0.7]

  rm(gene_sets_ordered)
  saveRDS(gene_sets_keep, file = rds)
}

gene_sets_sub <- gene_sets[names(gene_sets) %in% names(gene_sets_keep)]
gene_sets_sub
gene_sets_sub <- unique(unlist(geneIds(gene_sets_sub)))

gene_anno <- data.frame(
  symbol = unlist(seurat_orig@misc), ensembl = rownames(seurat_orig[["RNA"]])
)
rds <- file.path(cache_dir, "gene_anno.rds")
if (!file.exists(rds)) {
  saveRDS(gene_anno, file = rds)
}

ctypes <- vector("list", length = 7)
ctypes <- list(
  Astrocyte = c(
    "Astrocyte 01", "Astrocyte 02", "Astrocyte 03",
    "Astro Homeostatic 01", "Astro DAA 01"
  ),
  Oligodendrocyte = c(
    "Oligodendrocyte 01", "Oligodendrocyte 02", "Oligodendrocyte 03",
    "Oligodendrocyte 04", "Oligodendrocyte 05",
    "Oligo Homeostatic 01", "Oligo DAO 01"
  ),
  Microglia = c(
    "Microglia 01", "Microglia 02", "Micro Homeostatic 01",
    "Micro DAM 01", "Micro DAM 02"
  ),
  CA3 = c("Glut CA3 01", "Glut CA3 02", "Glut CA3 03"),
  CA1 = c("Glut CA1 01"),
  ENT = c("Glut L2 ENT 01", "Glut L3 ENT 01"),
  GABAergic = c("GABA Sst 01")
)

rds <- file.path(cache_dir, "seurat_list.rds")
if (file.exists(rds)) {
  rm(seurat_orig)
  gc()
  seurat_list <- readRDS(rds)
} else {

  seurat_list <- vector("list", length = 7)
  for (i in seq_along(ctypes)) {

    seurat <- subset(seurat_orig, idents = ctypes[[i]])

    seurat$genotype <- factor(
      seurat$genotype,
      levels = c(
        "MAPTKI", "S305N", "P301S", "NLGF_MAPTKI", "NLGF_S305N", "NLGF_P301S"
      )
    )
    seurat$genotype_batch <- factor(
      seurat$genotype_batch,
      levels = paste(
        rep(levels(seurat$genotype), each = 4),
        paste0("batch0", seq(4)), sep = "_"
      )
    )
    seurat$broad_annotations <- factor(seurat$broad_annotations)
    seurat$level1_subclusters <- factor(seurat$level1_subclusters)
    seurat$level2_subclusters <- factor(seurat$level2_subclusters)
    seurat@active.ident <- seurat$genotype_batch

    seurat$tau <- factor(
      gsub("NLGF_", replacement = "", x = seurat$genotype),
      levels = c("MAPTKI", "S305N", "P301S")
    )
    seurat$amyloid <- factor(
      ifelse(grepl("NLGF", x = seurat$genotype), yes = "NLGF", no = "WT"),
      levels = c("WT", "NLGF")
    )

    gc()
    DefaultAssay(seurat) <- "RNA"
    scores <- calculateScores(
      GetAssayData(seurat, layer = "counts"), rownames(seurat),
      geneIds(gene_sets)
    )
    gc()
    rownames(scores) <- names(gene_sets)
    colnames(scores) <- colnames(seurat)

    seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
    DefaultAssay(seurat) <- "GF"
    seurat$nCount_GF <- colSums(seurat, layer = "counts")
    seurat$nFeature_GF <- colSums(
      GetAssayData(seurat, layer = "counts") > 0
    )

    gc()
    DefaultAssay(seurat) <- "SCT"
    scores <- calculateScores(
      GetAssayData(seurat, layer = "counts"), rownames(seurat),
      geneIds(gene_sets)
    )
    gc()
    rownames(scores) <- names(gene_sets)
    colnames(scores) <- colnames(seurat)
    saveRDS(scores, file = rds)

    seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
    DefaultAssay(seurat) <- "GF_SCT"
    seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
    seurat$nFeature_GF_SCT <- colSums(
      GetAssayData(seurat, layer = "counts") > 0
    )

    seurat_list[[i]] <- seurat
  }
  rm(seurat_orig, seurat, scores)
  gc()
  saveRDS(seurat_list, file = rds)
}
seurat_list

gse_data <- vector("list", length = 7)
deg_data <- vector("list", length = 7)
gse_list <- vector("list", length = 7)
deg_list <- vector("list", length = 7)

for (i in seq_along(seurat_list)) {

  gse <- as.SingleCellExperiment(seurat_list[[i]], assay = "GF")
  gse <- gse[rownames(gse) %in% names(gene_sets_keep), ]
  gse <- suppressWarnings(
    aggregateAcrossCells(
      gse, colData(gse)[ , "genotype_batch"],
      use_exprs_values = "counts", statistics = "sum"
    )
  )

  deg <- as.SingleCellExperiment(seurat_list[[i]], assay = "RNA")
  deg <- deg[rownames(deg) %in% gene_sets_sub, ]
  deg <- suppressWarnings(
    aggregateAcrossCells(
      deg, colData(deg)[ , "genotype_batch"],
      use_exprs_values = "counts", statistics = "sum"
    )
  )

  dge_orig <- DGEList(counts(gse))
  keep <- filterByExpr(
    dge_orig, group = gse$tau,
    min.count = 10, min.total.count = 15
  )
  dge_orig <- dge_orig[keep, ]
  dge_orig <- calcNormFactors(dge_orig)
  dge <- DGEList(counts(gse))
  dge$samples$norm.factors <- dge_orig$samples$norm.factors
  keep <- filterByExpr(
    dge, group = gse$tau,
    min.count = 200, min.total.count = 300
  )
  dge <- dge[keep, , keep.lib.sizes = FALSE]
  dge <- calcNormFactors(dge)

  gse <- gse[rownames(gse) %in% rownames(dge)]
  logcounts(gse, withDimnames = FALSE) <- cpm(
    dge, log = TRUE, prior.count = 0.25
  )
  gse_data[[i]] <- gse

  dge_orig <- DGEList(counts(deg))
  keep <- filterByExpr(
    dge_orig, group = deg$tau,
    min.count = 10, min.total.count = 15
  )
  dge_orig <- dge_orig[keep, ]
  dge_orig <- calcNormFactors(dge_orig)
  dge <- DGEList(counts(deg))
  dge$samples$norm.factors <- dge_orig$samples$norm.factors
  keep <- filterByExpr(
    dge, group = deg$tau,
    min.count = 5, min.total.count = 7.5
  )
  dge <- dge[keep, , keep.lib.sizes = FALSE]
  dge <- calcNormFactors(dge)

  deg <- deg[rownames(deg) %in% rownames(dge)]
  logcounts(deg, withDimnames = FALSE) <- cpm(
    dge, log = TRUE, prior.count = 3
  )
  deg_data[[i]] <- deg

  design <- model.matrix(~ 0 + gse$genotype + gse$batch)
  colnames(design) <- make.names(colnames(design))
  cont_mat <- makeContrasts(
    S305N_MAPTKI =
      gse.genotypeS305N -
      gse.genotypeMAPTKI,
    P301S_MAPTKI =
      gse.genotypeP301S -
      gse.genotypeMAPTKI,
    NLGF_MAPTKI_MAPTKI =
      gse.genotypeNLGF_MAPTKI -
      gse.genotypeMAPTKI,
    NLGF_S305N_MAPTKI =
      gse.genotypeNLGF_S305N -
      gse.genotypeMAPTKI,
    NLGF_P301S_MAPTKI =
      gse.genotypeNLGF_P301S -
      gse.genotypeMAPTKI,
    levels = design
  )

  fit <- lmFit(logcounts(gse), design)
  fit <- contrasts.fit(fit, cont_mat)
  fit <- eBayes(fit, trend = TRUE, robust = TRUE)
  print(plotSA(fit))

  tests <- decideTests(fit, method = "global")
  write.fit(
    fit, tests, file.path(cache_dir, "results.tsv"),
    adjust = "BH", F.adjust = "BH", method = "global"
  )
  gse_list[[i]] <- read.delim(file.path(cache_dir, "results.tsv"))

  fit <- lmFit(logcounts(deg), design)
  fit <- contrasts.fit(fit, cont_mat)
  fit <- eBayes(fit, trend = TRUE, robust = TRUE)
  print(plotSA(fit))

  tests <- decideTests(fit, method = "global")
  write.fit(
    fit, tests, file.path(cache_dir, "results.tsv"),
    adjust = "BH", F.adjust = "BH", method = "global"
  )
  deg_list[[i]] <- read.delim(file.path(cache_dir, "results.tsv"))
}

results <- rbind(
  gse_list[[1]], deg_list[[1]], gse_list[[2]], deg_list[[2]],
  gse_list[[3]], deg_list[[3]], gse_list[[4]], deg_list[[4]],
  gse_list[[5]], deg_list[[5]], gse_list[[6]], deg_list[[6]],
  gse_list[[7]], deg_list[[7]]
)

corrections <- c(
  results[ , 13], results[ , 14], results[ , 15],
  results[ , 16], results[ , 17]
)
corrections <- data.frame(mean = rep(results[ , 2], times = 5), p = corrections)
corrections <- ihw(p ~ mean,  data = corrections, alpha = 0.05)
corrections <- adj_pvalues(corrections)
corrections <- split(
  corrections,
  f = gl(5, k = length(corrections) / 5, length = length(corrections))
)

rds <- file.path(cache_dir, "gse_data.rds")
if (!file.exists(rds)) {
  saveRDS(gse_data, file = rds)
}
rds <- file.path(cache_dir, "deg_data.rds")
if (!file.exists(rds)) {
  saveRDS(deg_data, file = rds)
}
rds <- file.path(cache_dir, "gse_list.rds")
if (!file.exists(rds)) {
  saveRDS(gse_list, file = rds)
}
rds <- file.path(cache_dir, "deg_list.rds")
if (!file.exists(rds)) {
  saveRDS(deg_list, file = rds)
}
rds <- file.path(cache_dir, "corrections.rds")
if (!file.exists(rds)) {
  saveRDS(corrections, file = rds)
}
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
