---
title: "Tau Mutant snRNAseq Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024-2025  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
conflicts_prefer(DT::JS)
packages <- c(
  "GSEABase", "SingleCellExperiment", "scales", "ComplexHeatmap", "circlize",
  "limma", "DT", "purrr", "dplyr", "networkD3"
)
lapply(packages, FUN = library, character.only = TRUE)

datatable_download_exp_15 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}
datatable_download_exp_7 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 7,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}

`%notin%` <- Negate(`%in%`)

data_dir <- file.path("..", "..", "cache", "comb", "step03")
gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_sets_keep <- readRDS(file.path(data_dir, "gene_sets_keep.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
deg_data <- readRDS(file.path(data_dir, "deg_data.rds"))
gse_data <- readRDS(file.path(data_dir, "gse_data.rds"))
deg_list <- readRDS(file.path(data_dir, "deg_list.rds"))
gse_list <- readRDS(file.path(data_dir, "gse_list.rds"))
gse_corrections <- readRDS(file.path(data_dir, "gse_corrections.rds"))
deg_corrections <- readRDS(file.path(data_dir, "deg_corrections.rds"))

map <- setNames(gene_anno$symbol, gene_anno$ensembl)

ctypes <- c(
  "Astrocytes", "Oligodendrocytes", "Microglia", "Glutamatergic CA3",
  "Glutamatergic CA1", "Glutamatergic L2/3 ENT", "GABAergic Sst"
)

anno_color <- hue_pal()(4)
top_anno <- HeatmapAnnotation(
  Batch = deg_data[[1]]$batch,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = sub(
      pattern = "_", replacement = " ", x = levels(deg_data[[1]]$genotype)
    ),
    labels_gp = gpar(col = "black", fontsize = 9),
    height = unit(5, "mm")
  ),
  col = list(
    Batch = c(
      "batch01" = anno_color[1], "batch02" = anno_color[2],
      "batch03" = anno_color[3], "batch04" = anno_color[4]
    )
  ),
  simple_anno_size = unit(3, "mm"),
  show_legend = c("Batch" = FALSE),
  annotation_name_gp = gpar(fontsize = 8)
)

deg <- deg_data[[1]]
design_deg <- model.matrix(~ 0 + deg$genotype + deg$batch)
colnames(design_deg) <- make.names(colnames(design_deg))
cont_mat_deg <- makeContrasts(
  S305N_MAPTKI =
    deg.genotypeS305N -
    deg.genotypeMAPTKI,
  P301S_MAPTKI =
    deg.genotypeP301S -
    deg.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    deg.genotypeNLGF_MAPTKI -
    deg.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    deg.genotypeNLGF_S305N -
    deg.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    deg.genotypeNLGF_P301S -
    deg.genotypeMAPTKI,
  levels = design_deg
)

gse <- gse_data[[1]]
design_gse <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design_gse) <- make.names(colnames(design_gse))
cont_mat_gse <- makeContrasts(
  S305N_MAPTKI =
    gse.genotypeS305N -
    gse.genotypeMAPTKI,
  P301S_MAPTKI =
    gse.genotypeP301S -
    gse.genotypeMAPTKI,
  NLGF_MAPTKI_MAPTKI =
    gse.genotypeNLGF_MAPTKI -
    gse.genotypeMAPTKI,
  NLGF_S305N_MAPTKI =
    gse.genotypeNLGF_S305N -
    gse.genotypeMAPTKI,
  NLGF_P301S_MAPTKI =
    gse.genotypeNLGF_P301S -
    gse.genotypeMAPTKI,
  levels = design_gse
)

ctype_name <- vector("list", length = 7)
deg_corrected <- vector("list", length = 7)
gse_corrected <- vector("list", length = 7)
deg_range <- vector("list", length = 7)
gse_range <- vector("list", length = 7)
results_deg <- vector("list", length = 7)
results_gse <- vector("list", length = 7)
sig_deg <- vector("list", length = 7)
sig_gse <- vector("list", length = 7)
sig_uniq_deg <- vector("list", length = 7)
sig_uniq_gse <- vector("list", length = 7)
```

Astrocytes
===

```{r, include = FALSE}
idx <- 1
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- 1
length2 <- nrow(deg_list[[1]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
colnames(sig_deg[[idx]]) <- colnames(deg_corrected[[idx]])
```

```{r}
checkboxGroupInput(
  paste0("genotype_deg", idx),
  label = "Select Genotypes:",
  selected = c(
    "MAPTKI", "S305N", "P301S", "NLGF_MAPTKI", "NLGF_S305N", "NLGF_P301S"
  ),
  inline = TRUE,
  choiceNames = c(
    "MAPTKI", "S305N", "P301S", "NLGF MAPTKI", "NLGF S305N", "NLGF P301S"
  ),
  choiceValues = c(
    "MAPTKI", "S305N", "P301S", "NLGF_MAPTKI", "NLGF_S305N", "NLGF_P301S"
  )
)

selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  multiple = TRUE,
  width = "100%"
)

observeEvent(
  input[[paste0("genotype_deg", 1)]], {
    sig_mat <- sig_deg[[1]][
      , sub("_batch.*", "", colnames(deg_corrected[[1]])) %in%
        input[[paste0("genotype_deg", 1)]]
    ]
    sig_mat <- sig_mat[
      apply(sig_mat, MARGIN = 1, FUN = function(x) any(x %in% c("↑", "↓"))), ,
      drop = FALSE
    ]
    if (nrow(sig_mat) > 30) {

      sel <- c()
      sig_mat_sub <- sig_mat
      while (length(sel) < 30) {
        sel <- c(sel, rownames(unique(sig_mat_sub)))
        remove <- which(rownames(sig_mat_sub) %in% sel)
        sig_mat_sub <- sig_mat_sub[-remove, ]
      }
      updateSelectizeInput(
        session,
        inputId = paste0("select_deg", 1),
        choices = rownames(sig_mat),
        selected = sel
      )

    } else {
      updateSelectizeInput(
        session,
        inputId = paste0("select_deg", 1),
        choices = rownames(sig_mat),
        selected = rownames(sig_mat)
      )
    }
  }
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- mat[
      , sub("_batch.*", "", colnames(deg_corrected[[idx]])) %in%
          input[[paste0("genotype_deg", idx)]],
      drop = FALSE
    ]

    top_anno2 <- HeatmapAnnotation(
      Batch = deg_data[[1]]$batch[
        sub("_batch.*", "", colnames(deg_corrected[[idx]])) %in%
          input[[paste0("genotype_deg", idx)]]
      ],
      cell_type = anno_block(
        gp = gpar(fill = "lightgray", col = "lightgray"),
        labels = sub(
          pattern = "_", replacement = " ",
          x = input[[paste0("genotype_deg", idx)]]
        ),
        labels_gp = gpar(col = "black", fontsize = 9),
        height = unit(5, "mm")
      ),
      col = list(
        Batch = c(
          "batch01" = anno_color[1], "batch02" = anno_color[2],
          "batch03" = anno_color[3], "batch04" = anno_color[4]
        )
      ),
      simple_anno_size = unit(3, "mm"),
      show_legend = c("Batch" = FALSE),
      annotation_name_gp = gpar(fontsize = 8)
    )

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(ncol(mat) / 4), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno2,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- 1
length2 <- nrow(gse_list[[1]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 1

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Oligodendrocytes
===

```{r, include = FALSE}
idx <- 2
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + 1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 2

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Microglia
===

```{r, include = FALSE}
idx <- 3
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + 1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 3

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Glutamatergic CA3
===

```{r, include = FALSE}
idx <- 4
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) + 1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 4
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 4

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Glutamatergic CA1
===

```{r, include = FALSE}
idx <- 5
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + 1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + nrow(deg_list[[5]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(gse_list[[5]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 5
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 5

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Glutamatergic L2/3 ENT
===

```{r, include = FALSE}
idx <- 6
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + nrow(deg_list[[5]]) + 1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + nrow(deg_list[[5]]) + nrow(deg_list[[6]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(gse_list[[5]]) + 1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(gse_list[[5]]) + nrow(gse_list[[6]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 6
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 6

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

GABAergic Sst
===

```{r, include = FALSE}
idx <- 7
ctype_name[[idx]] <- ctypes[idx]

deg <- deg_data[[idx]]
deg_corrected[[idx]] <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype
)
deg_range[[idx]] <- c(
  min(deg_corrected[[idx]]),
  (min(deg_corrected[[idx]]) + max(deg_corrected[[idx]])) / 2,
  max(deg_corrected[[idx]])
)

gse <- gse_data[[idx]]
gse_corrected[[idx]] <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype
)
gse_range[[idx]] <- c(
  min(gse_corrected[[idx]]),
  (min(gse_corrected[[idx]]) + max(gse_corrected[[idx]])) / 2,
  max(gse_corrected[[idx]])
)
```

Row {.tabset}
---

### Search Genes

```{r, include = FALSE}
fit <- lmFit(logcounts(deg), design_deg)
fit <- contrasts.fit(fit, cont_mat_deg)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_deg[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + nrow(deg_list[[5]]) + nrow(deg_list[[6]]) +1
length2 <- nrow(deg_list[[1]]) + nrow(deg_list[[2]]) + nrow(deg_list[[3]]) +
  nrow(deg_list[[4]]) + nrow(deg_list[[5]]) + nrow(deg_list[[6]]) +
  nrow(deg_list[[7]])
length <- length1:length2
for (i in seq_along(deg_corrections)) {
  results_deg[[idx]][ , col] <- deg_corrections[[i]][length]
  col <- col + 1
}

rownames(results_deg[[idx]]) <- results_deg[[idx]]$X
results_deg[[idx]] <- results_deg[[idx]][ , -1]
results_deg[[idx]] <- results_deg[[idx]][ , 1:23]
tests <- results_deg[[idx]][ , 17:21]
results_deg[[idx]] <- results_deg[[idx]][
  order(results_deg[[idx]]$F, decreasing = TRUE),
]
gene_anno_sub <- gene_anno[
  gene_anno$ensembl %in% rownames(results_deg[[idx]]),
]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg[[idx]]), gene_anno_sub$ensembl),
]
results_deg[[idx]]$Gene <- gene_anno_sub$symbol
results_deg[[idx]] <- results_deg[[idx]][
  , c(ncol(results_deg[[idx]]), 1:(ncol(results_deg[[idx]]) - 1))
]

sig_tmp <- ifelse(results_deg[[idx]][ , 18:22] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_deg[[idx]][ , 3:7] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_deg[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_deg[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_deg[[idx]] <- do.call(cbind, sig_deg[[idx]])
results_deg[[idx]] <- results_deg[[idx]][
  match(rownames(sig_deg[[idx]]), rownames(results_deg[[idx]])),
]
rownames(sig_deg[[idx]]) <- results_deg[[idx]]$Gene
sig_uniq_deg[[idx]] <- unique(sig_deg[[idx]])
```

```{r}
selectizeInput(
  paste0("select_deg", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("select_deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("select_deg", idx)]], gene_anno_sub$symbol),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(results_deg[[idx]])
```

### Search Gene Sets

```{r, include = FALSE}
fit <- lmFit(logcounts(gse), design_gse)
fit <- contrasts.fit(fit, cont_mat_gse)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(
  fit, tests, file = tc,
  adjust = "BH", F.adjust = "BH", method = "global"
)
close(tc)
results_gse[[idx]] <- read.delim(text = results)

col <- 18
length1 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(gse_list[[5]]) + nrow(gse_list[[6]]) +1
length2 <- nrow(gse_list[[1]]) + nrow(gse_list[[2]]) + nrow(gse_list[[3]]) +
  nrow(gse_list[[4]]) + nrow(gse_list[[5]]) + nrow(gse_list[[6]]) +
  nrow(gse_list[[7]])
length <- length1:length2
for (i in seq_along(gse_corrections)) {
  results_gse[[idx]][ , col] <- gse_corrections[[i]][length]
  col <- col + 1
}

rownames(results_gse[[idx]]) <- results_gse[[idx]]$X
results_gse[[idx]] <- results_gse[[idx]][ , -1]
results_gse[[idx]] <- results_gse[[idx]][ , 1:23]
tests <- results_gse[[idx]][ , 17:21]
results_gse[[idx]] <- results_gse[[idx]][
  order(results_gse[[idx]]$F, decreasing = TRUE),
]
results_gse[[idx]]$GeneSet <- rownames(results_gse[[idx]])

sig_tmp <- ifelse(results_gse[[idx]][ , 17:21] < 0.05, yes = 1, no = 0)
direction <- ifelse(results_gse[[idx]][ , 2:6] > 0, yes = 1, no = -1)
sig_tmp <- sig_tmp * direction

row_scores <- t(
  apply(
    sig_tmp,
    MARGIN = 1,
    function(row) {
      NLGF_P301S_up <- sum(row[5] == 1) * 5
      NLGF_S305N_up <- sum(row[4] == 1) * 4
      NLGF_MAPTKI_up <- sum(row[3] == 1) * 3
      P301S_up <- sum(row[2] == 1) * 2
      S305N_up <- sum(row[1] == 1) * 1
      S305N_dn <- sum(row[1] == -1) * -1
      P301S_dn <- sum(row[2] == -1) * -2
      NLGF_MAPTKI_dn <- sum(row[3] == -1) * -3
      NLGF_S305N_dn <- sum(row[4] == -1) * -4
      NLGF_P301S_dn <- sum(row[5] == -1) * -5
      return(
        c(
          NLGF_P301S_up, NLGF_S305N_up, NLGF_MAPTKI_up, P301S_up, S305N_up,
          NLGF_P301S_dn, NLGF_S305N_dn, NLGF_MAPTKI_dn, P301S_dn, S305N_dn
        )
      )
    }
  )
)
order <- order(
  row_scores[, 1], row_scores[, 2], row_scores[, 3], row_scores[, 4],
  row_scores[, 5], row_scores[, 6], row_scores[, 7], row_scores[, 8],
  row_scores[, 9], row_scores[, 10], decreasing = TRUE
)
sig_tmp <- sig_tmp[order, ]

sig_tmp <- ifelse(
  sig_tmp == -1, yes = "↓", no = ifelse(sig_tmp == 1, yes = "↑", no = "")
)
sig_tmp <- cbind(rep("", nrow(sig_tmp)), sig_tmp)
sig_gse[[idx]] <- vector("list", length = 6)
for (i in seq_len(ncol(sig_tmp))) {
  sig_gse[[idx]][[i]] <- replicate(4, expr = sig_tmp[ , i])
}
sig_gse[[idx]] <- do.call(cbind, sig_gse[[idx]])
results_gse[[idx]] <- results_gse[[idx]][
  match(rownames(sig_gse[[idx]]), rownames(results_gse[[idx]])),
]
sig_uniq_gse[[idx]] <- unique(sig_gse[[idx]])
```

```{r}
selectizeInput(
  paste0("select_gse", idx),
  label = "Select Gene Sets:",
  choices = results_gse[[idx]]$GeneSet,
  selected = rownames(sig_uniq_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    sig_tmp <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("select_gse", idx)]], ,
        drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(input[[paste0("select_gse", idx)]], rownames(sig_tmp)),
      ]
    }

    mat <- gse_corrected[[idx]][
      rownames(gse_corrected[[idx]]) %in% rownames(sig_tmp), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig_tmp), rownames(mat)), ]
    }

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = ctype_name[[idx]],
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        row_names_max_width = max_text_width(rownames(mat)),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logGSE", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 650,
  res = 96
)
```

</div>
</div>

### Gene Set Stats

```{r}
datatable_download_exp_7(results_gse[[idx]])
```

### Gene Set Genes

```{r}
selectizeInput(
  paste0("select_gse_genes", idx),
  label = "Select Gene Sets:",
  choices = names(gene_sets[names(gene_sets) %in% names(gene_sets_keep)]),
  selected = rownames(results_gse[[idx]])[which.max(results_gse[[idx]]$F)],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(deg_range[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 7
    genes <- unlist(
      geneIds(
        gene_sets[
          names(gene_sets) %in% input[[paste0("select_gse_genes", idx)]]
        ]
      )
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig_tmp <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig_tmp) > 1) {
      sig_tmp <- sig_tmp[
        match(gene_anno_sub$symbol, rownames(sig_tmp)), , drop = FALSE
      ]
    }

    mat <- deg_corrected[[idx]][
      rownames(deg_corrected[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = 24)
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig_tmp <- rbind(sig_tmp, add)
    sig_tmp[is.na(sig_tmp)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          ctype_name[[idx]],
          ": ",
          gene_sets[[
            input[[paste0("select_gse_genes", idx)]]
          ]]@shortDescription,
          "\n",
          input[[paste0("select_gse_genes", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(6), each = 4),
        column_gap = unit(1.5, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 10),
        row_title_rot = 0,
        top_annotation = top_anno,
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig_tmp[i, j], x = x, y = y, gp = gpar(fontsize = 10))
        },
        heatmap_legend_param = list(
          title = "Scaled logCPM", direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 8),
          grid_height = unit(2, "mm"), title_gp = gpar(fontsize = 10)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Connectivity

```{r}
selectizeInput(
  paste0("deg_conn", idx),
  label = "Select Genes:",
  choices = results_deg[[idx]]$Gene,
  selected = rownames(sig_uniq_deg[[idx]])[1:2],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork(
  {
    idx <- 7

    links_list <- vector(
      "list", length = length(input[[paste0("deg_conn", idx)]])
    )
    for (i in seq_along(input[[paste0("deg_conn", idx)]])) {
      gse_sub <- results_gse[[idx]][results_gse[[idx]]$F.p.value < 0.05, ]
      gene_ids <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub$GeneSet]
      gene_ids <- gene_ids[order(names(gene_ids))]

      links_orig <- data.frame(
        source = rep(names(gene_ids), times = lengths(gene_ids)),
        target = gene_ids %>% map_dfr(as_tibble)
      )
      links <- links_orig
      links_orig[] <- map[unlist(links_orig)]
      links$value <- links_orig$value
      colnames(links) <- c("source", "target")
      links <- links[links$target %in% input[[paste0("deg_conn", idx)]][i], ]
      links$value <- rep(1, times = nrow(links))
      links_list[[i]] <- links
    }
    links_comb <- do.call(rbind, args = links_list)

    nodes <- data.frame(
      name = c(
        as.character(links_comb$source), as.character(links_comb$target)
      ) %>% unique()
    )
    links_comb$IDsource <- match(links_comb$source, nodes$name) - 1
    links_comb$IDtarget <- match(links_comb$target, nodes$name) - 1

    if (nrow(nodes) > 0) {
      select_length <- length(
        which(input[[paste0("deg_conn", idx)]] %in% nodes$name)
      )
      nodes$group <- c(
        rep("target", times = nrow(nodes) - select_length),
        nodes$name[(nrow(nodes) - (select_length - 1)):nrow(nodes)]
      )
      colour <- c(hue_pal()(select_length), "#D3D3D3")
      colour <- paste0(
        'd3.scaleOrdinal().range(["', paste(colour, collapse = '", "'), '"])'
      )

      sankeyNetwork(
        Links = links_comb, Nodes = nodes, Source = "IDsource",
        Target = "IDtarget", Value = "value", NodeID = "name",
        sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
        LinkGroup = "target", colourScale = colour, margin = list(bottom = 100)
      )
    }
  }
)
```

Viewer Documentation
===

<div style='font-size: 18px; padding-left: 500px; padding-right: 500px;'>

#### Search Genes

In this section, the user can create heatmaps for genes of interest.  
By default, it is loaded to show all unique combinations of significant genes.  
The heatmap on the right shows scaled expression values, where in each row, the lowest value is assigned -1 and the highest value is assigned 1.  
The heatmap on the left shows the "real" log counts-per-million values without scaling.  
A cell is assigned and up or down arrow if there is a significant change (adjusted P-value < 0.05) in the corresponding direction compared to the MAPTKI group.  
Users can add or remove genes as desired and the heatmap will update.

#### Gene Stats

Stats tables for all genes that were tested, note that no significance cutoffs are applied in order to show all genes.  
Users can search for genes, sort columns, change the number of rows in view, and download the table.

#### Search Gene Sets

This section is similar to *Search Genes*, except shown and tested are gene sets from the GO Ontology.  
A custom gene set enrichment method is used to aggregate genes into the gene sets.  
Only the scaled heatmap is shown due to the space taken by gene set label names.

#### Gene Set Stats

This section is similar to *Gene Stats*, but for the gene sets.

#### Gene Set Genes

Here users can input a gene set that was tested and see the expression of the genes within the gene set.  
The output is similar to *Search Genes*.  
Genes that which expression is too low to be quantified have cells shaded in gray.

#### Gene Connectivity

Here users can input any number of genes and visualise the genes that they are a part of.  
Gene sets are filtered to those that are likely significant in at least one comparison (F P-value < 0.05), but currently does not signify which comparison.

</div>

```{r, include = FALSE}
rm(
  deg, deg_data, deg_list,
  gse, gse_data, gse_list,
  deg_corrections, gse_corrections
)
gc()
```
