---
title: "Combined Samples - 03 GeneFunnel"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "results", "comb", "comb_03genefunnel.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-snRNAseq](https://github.com/eturkes/tau-mutant-snRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(
  AnnotationDbi::select, edgeR::cpm, DT::JS,
  base::intersect, reshape2::melt, .quiet = TRUE
)
packages <- c(
  "GSEABase", "Seurat", "ggplot2", "dplyr", "ggrepel", "scuttle", "GO.db", "tm",
  "pals", "ComplexHeatmap", "edgeR", "DT", "circlize", "purrr", "networkD3",
  "factoextra", "reshape2", "dendextend", "data.table", "scales", "igraph",
  "qgraph", "multienrichjam", "GSVA", "Rcpp", "biomaRt"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 96)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
sourceCpp(file.path("..", "..", "src", "calculateScores.cpp"))

`%notin%` <- Negate(`%in%`)
# -----------------

# Useful variables.
# -----------------
step <- 3

cache_dir <- file.path("..", "..", "cache", "comb", paste0("step", "0", step))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------
```

# Prep

```{r}
rds <- file.path(cache_dir, "seurat_orig.rds")
if (file.exists(rds)) {
  seurat_orig <- readRDS(rds)
} else {

  file_list <- list.files(file.path(cache_dir, "..", ".."), recursive = TRUE)
  file_list <- file_list[grep("seurat_prep.rds", file_list)]
  files <- vector("list", length(file_list))
  for (i in seq_along(files)) {
    files[[i]] <- readRDS(file.path(cache_dir, "..", "..", file_list[i]))
    files[[i]][["RNA"]] <- NULL
    files[[i]]@meta.data[grep("SCT", colnames(files[[i]]@meta.data))] <- NULL
  }

  gc()
  seurat_orig <- merge(
    files[[1]], c(files[[2]], files[[3]], files[[4]],
    files[[5]], files[[6]], files[[7]], files[[8]],
    files[[9]], files[[10]], files[[11]], files[[12]],
    files[[13]], files[[14]], files[[15]], files[[16]],
    files[[17]], files[[18]], files[[19]], files[[20]],
    files[[21]], files[[22]], files[[23]], files[[24]]
    )
  )
  rm(files)
  gc()

  seurat_anno <- readRDS(
    file.path(cache_dir, "..", "step02", "annotated_seurat.rds")
  )
  seurat_orig[["RNA"]] <- CreateAssay5Object(
    GetAssayData(seurat_anno, assay = "RNA")
  )
  seurat_orig@meta.data <- seurat_anno[[]]
  seurat_orig[["pca"]] <- seurat_anno[["pca"]]
  seurat_orig[["umap"]] <- seurat_anno[["umap"]]
  seurat_orig@misc <- seurat_anno@misc
  seurat_orig@active.ident <- seurat_orig$genotype_batch

  rm(seurat_anno)
  gc()
  saveRDS(seurat_orig, file = rds)
}
seurat_orig

rds <- file.path(cache_dir, "filtered_gene_sets_mmusculus.rds")
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {

  gene_sets <- getGmt(
    file.path(
      "..", "..", "data",
      "gprofiler_mmusculus_ENSG_2024_06_15", "mmusculus.GO.comb.ENSG.gmt"
    )
  )
  keep <- filterGeneSets(gene_sets, min.sz = 2)
  gene_sets <- gene_sets[names(gene_sets) %in% names(keep)]

  remove <- unique(unlist(geneIds(gene_sets)))[
    unique(unlist(geneIds(gene_sets))) %notin% rownames(seurat_orig)
  ]
  remove <- which(
    sapply(geneIds(gene_sets), FUN = function(x) any(x %in% remove))
  )
  gene_sets <- gene_sets[-remove]

  for (i in seq_along(gene_sets@.Data)) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets, uniqGenes = unique(unlist(geneIds(gene_sets)))
  )
  gene_sets_keep <- rowSums(gene_sets_keep)
  gene_sets_keep <- gene_sets_keep[order(gene_sets_keep, decreasing = TRUE)]
  gene_sets_keep <- gene_sets[match(names(gene_sets_keep), names(gene_sets))]

  dup <- sapply(
    geneIds(gene_sets_keep), FUN = function(x) paste(sort(x), collapse = "-")
  )
  dup <- which(duplicated(dup))
  gene_sets <- gene_sets_keep[-dup]

  saveRDS(gene_sets, file = rds)
}
gene_sets

rds <- file.path(cache_dir, "gene_sets_keep.rds")
if (file.exists(rds)) {
  gene_sets_keep <- readRDS(rds)
} else {

  mart <- useEnsembl("ensembl", "mmusculus_gene_ensembl")
  attributes <- c("ensembl_gene_id", "chromosome_name")
  gene_anno <- getBM(
    attributes, filters = "ensembl_gene_id",
    values = unique(unlist(geneIds(gene_sets))), mart = mart
  )
  remove <- gene_anno[
    gene_anno$chromosome_name == "Y" | gene_anno$chromosome_name == "X",
  ]$ensembl_gene_id
  remove <- which(
    sapply(geneIds(gene_sets), FUN = function(x) any(x %in% remove))
  )
  gene_sets_keep <- gene_sets[-remove]

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets_keep, uniqGenes = unique(unlist(geneIds(gene_sets_keep)))
  )
  gene_sets_keep[upper.tri(gene_sets_keep)] <- 0
  diag(gene_sets_keep) <- 0
  gene_sets_keep <- apply(gene_sets_keep, MARGIN = 1, FUN = max)
  gene_sets_keep <- gene_sets_keep[gene_sets_keep <= 0.25]

  rm(mart)
  saveRDS(gene_sets_keep, file = rds)
}
gene_sets[names(gene_sets) %in% names(gene_sets_keep)]

gene_anno <- data.frame(
  geneid = unlist(seurat_orig@misc), gene = rownames(seurat_orig[["RNA"]])
)
```

# Glut L2/3 Cortex 01

```{r}
ctype <- "Glut L2-3 Cortex 01"
```

## MAPT KI vs. S305N+3

```{r}
genotype <- c("MAPTKI", "S305N")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "S305N"),
    paste0("gse.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "S305N"),
    paste0("counts.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## MAPT KI vs. P301S+3

```{r}
genotype <- c("MAPTKI", "P301S")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "P301S"),
    paste0("gse.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "P301S"),
    paste0("counts.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## S305N+3 vs. P301S+3

```{r}
genotype <- c("S305N", "P301S")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "P301S"),
    paste0("gse.genotype", "S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "P301S"),
    paste0("counts.genotype", "S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## MAPT KI vs. NLGF MAPT KI

```{r}
genotype <- c("MAPTKI", "NLGF_MAPTKI")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_MAPTKI"),
    paste0("gse.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_MAPTKI"),
    paste0("counts.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## S305N+3 vs. NLGF S305N+3

```{r}
genotype <- c("S305N", "NLGF_S305N")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_S305N"),
    paste0("gse.genotype", "S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_S305N"),
    paste0("counts.genotype", "S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## P301S+3 vs. NLGF P301S+3

```{r}
genotype <- c("P301S", "NLGF_P301S")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_P301S"),
    paste0("gse.genotype", "P301S"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_P301S"),
    paste0("counts.genotype", "P301S"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## NLGF S305N+3 vs. NLGF P301S+3

```{r}
genotype <- c("NLGF_S305N", "NLGF_P301S")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_P301S"),
    paste0("gse.genotype", "NLGF_S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_P301S"),
    paste0("counts.genotype", "NLGF_S305N"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## MAPT KI vs. NLGF S305N+3

```{r}
genotype <- c("MAPTKI", "NLGF_S305N")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_S305N"),
    paste0("gse.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_S305N"),
    paste0("counts.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## MAPT KI vs. NLGF P301S+3

```{r}
genotype <- c("MAPTKI", "NLGF_P301S")

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat_cleaned.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  seurat <- seurat_orig
  seurat@active.ident <- seurat$level2_subclusters
  seurat <- subset(seurat, idents = gsub("-", replacement = "/", x = ctype))
  seurat@active.ident <- seurat$genotype
  seurat <- subset(seurat, idents = genotype)
  seurat@active.ident <- factor(seurat$genotype_batch)
  seurat <- subset(seurat, downsample = min(table(seurat@active.ident)))

  seurat <- PrepSCTFindMarkers(seurat, verbose = FALSE)

  counts_mat <- GetAssayData(seurat)
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, new.data = counts_mat)
  counts_mat <- GetAssayData(seurat, slot = "counts")
  counts_mat[is.na(counts_mat)] <- 0
  seurat <- SetAssayData(seurat, slot = "counts", new.data = counts_mat)

  seurat$nCount_SCT <- colSums(seurat, layer = "counts")
  seurat$nFeature_SCT <- colSums(counts_mat > 0)

  seurat$genotype <- factor(seurat$genotype, levels = c(genotype))
  seurat$genotype_batch <- factor(
    seurat$genotype_batch,
    levels = paste(
      rep(levels(seurat$genotype), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat$broad_annotations <- factor(seurat$broad_annotations)
  seurat$level1_subclusters <- factor(seurat$level1_subclusters)
  seurat$level2_subclusters <- factor(seurat$level2_subclusters)
  seurat@active.ident <- seurat$genotype_batch

  saveRDS(seurat, file = rds)
}

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "RNA"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF"
seurat$nCount_GF <- colSums(seurat, layer = "counts")
seurat$nFeature_GF <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "scores_SCT.rds",
    sep = "_"
  )
)
if (file.exists(rds)) {
  scores <- readRDS(rds)
} else {
  gc()
  DefaultAssay(seurat) <- "SCT"
  scores <- calculateScores(
    GetAssayData(seurat, layer = "counts"), rownames(seurat), geneIds(gene_sets)
  )
  gc()
  rownames(scores) <- names(gene_sets)
  colnames(scores) <- colnames(seurat)
  saveRDS(scores, file = rds)
}
seurat[["GF_SCT"]] <- CreateAssay5Object(scores, data = log1p(scores))
DefaultAssay(seurat) <- "GF_SCT"
seurat$nCount_GF_SCT <- colSums(seurat, layer = "counts")
seurat$nFeature_GF_SCT <- colSums(GetAssayData(seurat, layer = "counts") > 0)

rm(scores)
rds <- file.path(
  cache_dir,
  paste(
    gsub(" ", "_", x = paste(ctype, collapse = " ")),
    gsub(" ", "_", x = paste(genotype, collapse = " ")),
    "seurat.rds",
    sep = "_"
  )
)
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
seurat
```

```{r}
thresh <- table(seurat$genotype_batch)
thresh
thresh <- unique(thresh) * 0.01
thresh

gse <- as.SingleCellExperiment(seurat, assay = "GF_SCT")
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(gse), group = factor(gse$genotype),
  min.total.count = thresh * 2, min.count = thresh * 2
)
gse <- gse[keep, ]
logcounts(gse, withDimnames = FALSE) <- cpm(
  counts(gse), log = TRUE, prior.count = 1
)
gse

counts <- as.SingleCellExperiment(seurat, assay = "SCT")
counts <- suppressWarnings(
  aggregateAcrossCells(
    counts, colData(counts)[ , "genotype_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
keep <- filterByExpr(
  counts(counts), group = factor(counts$genotype),
  min.total.count = thresh, min.count = thresh
)
counts <- counts[keep, ]
logcounts(counts, withDimnames = FALSE) <- cpm(
  counts(counts), log = TRUE, prior.count = 2
)
counts

gse_cpm <- logcounts(gse)
counts_cpm <- logcounts(counts)
gse_range <- c(
  min(gse_cpm), (min(gse_cpm) + max(gse_cpm)) / 2, max(gse_cpm)
)
counts_range <- c(
  min(counts_cpm), (min(counts_cpm) + max(counts_cpm)) / 2, max(counts_cpm)
)
```

```{r}
design <- model.matrix(~ 0 + gse$genotype + gse$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(gse_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("gse.genotype", "NLGF_P301S"),
    paste0("gse.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[rownames(cont_fit) %in% names(gene_sets_keep), ]
gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(cont_fit)]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)
gse_results_up <- results[results$Coef > 0, ]
gse_results_down <- results[results$Coef < 0, ]

datatable_download(gse_results_up)
datatable_download(gse_results_down)
```

```{r, fig.width = 14, fig.height = 8}
mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_up)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_cpm[
  rownames(gse_cpm) %in% rownames(gse_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(gse_results_down)[1:30], rownames(mat)), , drop = FALSE
]
mat <- removeBatchEffect(mat, batch = gse$batch, group = gse$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

```{r}
design <- model.matrix(~ 0 + counts$genotype + counts$batch)
colnames(design) <- make.names(colnames(design))

fit <- lmFit(counts_cpm, design)
cont_mat <- makeContrasts(
  paste(
    paste0("counts.genotype", "NLGF_P301S"),
    paste0("counts.genotype", "MAPTKI"),
    sep = "-"
  ),
  levels = design
)
cont_mat <- contrasts.fit(fit, cont_mat)
cont_fit <- eBayes(cont_mat, trend = TRUE, robust = TRUE)
plotSA(cont_fit)
cont_fit <- cont_fit[
  rownames(cont_fit) %in% unique(unlist(geneIds(gene_sets_sub))),
]
tests <- decideTests(cont_fit)
write.fit(
  cont_fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))
rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$gene %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$gene),
]
results$gene <- gene_anno_sub$geneid
deg_results_up <- results[results$Coef > 0, ]
deg_results_down <- results[results$Coef < 0, ]

datatable_download(deg_results_up)
datatable_download(deg_results_down)
```

```{r, fig.width = 7, fig.height = 8}
mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_up$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- counts_cpm[
  rownames(counts_cpm) %in% rownames(deg_results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(deg_results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- deg_results_down$gene[1:30]
mat <- removeBatchEffect(mat, batch = counts$batch, group = counts$genotype)
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
