---
title: "NLGF S305N+3 - Batch 03 - 02 Cell-typing"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "results", "batch03", "NLGF_S305N_batch03_02celltyping.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-snRNAseq](https://github.com/eturkes/tau-mutant-snRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(DT::JS, .quiet = TRUE)
packages <- c("Seurat", "parallelly", "ggplot2", "dplyr", "ggrepel", "DT")
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 300)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
# -----------------

# Useful variables.
# -----------------
genotype <- "NLGF_S305N"
batch <- 3
step <- 2

data_dir <- file.path(
  "..", "..", "data", "external", "tau-mutant-snRNAseq",
  "cellranger", "batch3_4", "NWB3_6", "outs", "raw_feature_bc_matrix"
)
cache_dir <- file.path(
  "..", "..", "cache",
  paste0("batch", "0", batch), genotype, paste0("step", "0", step)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------
```

# Cell-typing

We read in the previous Seurat object and annotate the clusters, using the Allen Brain Atlas as a reference.

```{r}
seurat <- readRDS(file.path(cache_dir, "..", "step01", "seurat_prep.rds"))
seurat

geneids <- read.delim(file.path(data_dir, "features.tsv.gz"), header = FALSE)
geneids <- geneids[geneids$V1 %in% rownames(seurat), ]
dup <- geneids[
  duplicated(geneids$V2) | duplicated(geneids$V2, fromLast = TRUE),
]
for (i in seq(nrow(dup))) {
  for (j in seq(nrow(geneids))) {
    if (dup$V1[i] == geneids$V1[j]) {
      geneids$V2[j] <- paste(geneids$V2[j], geneids$V1[j], sep = "-")
    }
  }
}
seurat@misc <- list(geneids = geneids$V2)
rm(geneids)

seurat_geneids <- GetAssayData(seurat)
rownames(seurat_geneids) <- seurat@misc$geneids
seurat_geneids <- CreateSeuratObject(
  counts = seurat_geneids, data = seurat_geneids,
  assay = "SCT", meta.data = seurat@meta.data
)
seurat_geneids[["pca"]] <- seurat[["pca"]]
seurat_geneids[["umap"]] <- seurat[["umap"]]
seurat_geneids@active.ident <- seurat_geneids$seurat_clusters

rds <- file.path(cache_dir, "markers.rds")
if (file.exists(rds)) {
  markers <- readRDS(rds)
} else {
  markers <- FindAllMarkers(seurat_geneids, only.pos = TRUE, verbose = FALSE)
  saveRDS(markers, file = rds)
}

markers <- markers[markers$p_val_adj < 0.05, ]
markers <- markers %>% distinct(gene, .keep_all = TRUE)
top <- markers %>% group_by(cluster) %>% top_n(n = -25, wt = p_val_adj)
datatable_download(top)

rds <- file.path(cache_dir, "allen_labels.rds")
if (file.exists(rds)) {
  allen_labels <- readRDS(rds)
} else {
  ref <- readRDS(
    file.path(
      cache_dir, "..", "..", "..", "misc", "allen_mouse_hip_ctx_10x.rds"
    )
  )
  anchors <- FindTransferAnchors(
    ref, seurat_geneids, normalization.method = "SCT",
    recompute.residuals = FALSE, reference.reduction = "pca", dims = 1:30
  )
  predictions <- MapQuery(
    anchors, seurat_geneids, ref,
    refdata = "subclass_label", reference.reduction = "umap"
  )
  allen_labels <- predictions$predicted.id
  rm(ref, anchors, predictions)
  saveRDS(allen_labels, file = rds)
}

seurat$allen_labels <- allen_labels
red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "allen_labels", type = "cat"
)
red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "seurat_clusters", type = "cat"
)

table <- table(seurat$seurat_clusters, seurat$allen_labels)
max <- colnames(table)[apply(table, MARGIN = 1, FUN = which.max)]
max

seurat <- RenameIdents(
  object = seurat, "0" = "Oligodendrocyte",
  "1" = "Microglia", "2" = "Cortical Glutamatergic",
  "3" = "Hippocampal Glutamatergic", "4" = "Astrocyte",
  "5" = "Cortical Glutamatergic", "6" = "Hippocampal Glutamatergic",
  "7" = "Hippocampal Glutamatergic", "8" = "Cortical Glutamatergic",
  "9" = "Oligodendrocyte", "10" = "Vascular",
  "11" = "Cortical Glutamatergic", "12" = "OPC",
  "13" = "GABAergic", "14" = "Cortical Glutamatergic",
  "15" = "GABAergic", "16" = "Unknown",
  "17" = "Cortical Glutamatergic", "18" = "Vascular",
  "19" = "Cortical Glutamatergic", "20" = "Cortical Glutamatergic",
  "21" = "Microglia", "22" = "Unknown",
  "23" = "Cortical Glutamatergic", "24" = "Cortical Glutamatergic",
  "25" = "Unknown"
)
seurat@active.ident <- factor(as.character(seurat@active.ident))
seurat$broad_annotations <- seurat@active.ident
red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "broad_annotations", type = "cat"
)

rds <- file.path(cache_dir, "seurat_annotated.rds")
if (!file.exists(rds)) {
  saveRDS(seurat, file = rds)
}
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
