---
title: "NLGF P301S+3 - Batch 04 - 01 Prep"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "results", "batch4", "NLGF_P301S_batch04_01prep.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-snRNAseq](https://github.com/eturkes/tau-mutant-snRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-snRNAseq.
#    Copyright (C) 2024  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
packages <- c(
  "Seurat", "parallelly", "ggplot2", "dplyr", "ggrepel", "viridis",
  "DropletUtils", "scDblFinder", "BiocParallel", "DT", "biomaRt"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 300)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
# -----------------

# Useful variables.
# -----------------
genotype <- "NLGF_P301S"
batch <- 4
step <- 1

data_dir <- file.path(
  "..", "..", "data", "external", "tau-mutant-snRNAseq",
  "cellranger", "batch3_4", "NWB4_4", "outs", "raw_feature_bc_matrix"
)
cache_dir <- file.path(
  "..", "..", "cache",
  paste0("batch", "0", batch), genotype, paste0("step", "0", step)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

protocol <- c("mouse", "droplet", "single-nuc", "umis")
vars_to_regress <- NULL
parallel_override <- NULL
# -----------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", length = 1)
names(metadata_to_plot) <- "seurat_clusters"
# -------------------------------------------------------------------------
```

# Cleaning

We first read in the data and form a Seurat object.

```{r}
seurat <- CreateSeuratObject(Read10X(data_dir, gene.column = 1))
seurat
```

# QC

In a droplet-based experiment, we first remove empty droplets.

```{r}
if (protocol[2] == "droplet") {

  rds <- file.path(cache_dir, "empty_drops.rds")
  if (file.exists(rds)) {
    empty_drops <- readRDS(rds)
  } else {
    set.seed(1)
    empty_drops <- emptyDrops(
      GetAssayData(seurat), BPPARAM = MulticoreParam(availableCores()),
      lower = 20
    )
    saveRDS(empty_drops, file = rds)
  }

  bcrank <- barcodeRanks(GetAssayData(seurat), lower = 20)
  uniq <- !duplicated(bcrank$rank)
  plot(
    bcrank$rank[uniq], bcrank$total[uniq], log = "xy",
    xlab = "Rank", ylab = "Total UMI count", cex.lab = 1.2
  )
  abline(h = metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
  abline(h = metadata(bcrank)$knee, col = "dodgerblue", lty = 2)
  legend(
    "bottomleft", legend = c("Inflection", "Knee"),
    col = c("darkgreen", "dodgerblue"),
    lty = 2, cex = 1.2
  )

  seurat <- seurat[ , which(empty_drops$FDR <= 0.001)]
  seurat

  bcrank <- barcodeRanks(GetAssayData(seurat), lower = 20)
  uniq <- !duplicated(bcrank$rank)
  plot(
    bcrank$rank[uniq], bcrank$total[uniq], log = "xy",
    xlab = "Rank", ylab = "Total UMI count", cex.lab = 1.2
  )
  abline(h = metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
  abline(h = metadata(bcrank)$knee, col = "dodgerblue", lty = 2)
  legend(
    "bottomleft", legend = c("Inflection", "Knee"),
    col = c("darkgreen", "dodgerblue"),
    lty = 2, cex = 1.2
  )

  rm(bcrank, empty_drops)
}
```

We remove cells low in unique features and high in mito/ribo reads.
Mitochondrial reads themselves are also removed in a snRNAseq experiment.

```{r}
mart <- useEnsembl("ensembl", dataset = "mmusculus_gene_ensembl")
attributes <- c("external_gene_name", "ensembl_gene_id")
gene_anno <- getBM(
  attributes, filters = "ensembl_gene_id",
  values = rownames(seurat), mart = mart
)

mito <- grep("^mt-", gene_anno$external_gene_name)
mito <- which(rownames(seurat) %in% gene_anno[mito, ]$ensembl_gene_id)
seurat[["percent_mt"]] <- PercentageFeatureSet(
  seurat, features = rownames(seurat)[mito]
)
ribo <- grep("^Rp[sl][[:digit:]]", gene_anno$external_gene_name)
ribo <- which(rownames(seurat) %in% gene_anno[ribo, ]$ensembl_gene_id)
seurat[["percent_ribo"]] <- PercentageFeatureSet(
  seurat, features = rownames(seurat)[ribo]
)

if (length(mito) != 0) {
  if (protocol[3] == "single-nuc") {
    cat(paste0(
      "Removed ", length(mito), " mitochondrial genes:"), rownames(seurat)[mito]
    )
    seurat <- seurat[-mito, ]
    seurat
  } else {
    cat(
      paste0(
        "Found but did not remove ", length(mito), " mitochondrial genes:"
      ),
      rownames(seurat)[mito]
    )
  }
}

VlnPlot(
  seurat,
  features = c("nFeature_RNA", "nCount_RNA", "percent_mt", "percent_ribo"),
  ncol = 4
)

seurat@meta.data %>%
  ggplot(aes(x = nFeature_RNA)) +
  geom_density(alpha = 0.2) +
  scale_x_log10() +
  theme_classic() +
  ylab("Cell Density") +
  geom_vline(xintercept = 500)

seurat@meta.data %>%
  ggplot(aes(x = nCount_RNA)) +
  geom_density(alpha = 0.2) +
  scale_x_log10() +
  theme_classic() +
  ylab("Cell Density") +
  geom_vline(xintercept = 600)

seurat <- subset(
  seurat,
  subset = nFeature_RNA >= 400 & nCount_RNA >= 500 & nFeature_RNA < 6000 &
    nCount_RNA < 20000 & percent_mt < 5 & percent_ribo < 10
)
seurat

VlnPlot(
  seurat,
  features = c("nFeature_RNA", "nCount_RNA", "percent_mt", "percent_ribo"),
  ncol = 4
)

seurat@meta.data %>%
  ggplot(aes(x = nFeature_RNA)) +
  geom_density(alpha = 0.2) +
  scale_x_log10() +
  theme_classic() +
  ylab("Cell Density")

seurat@meta.data %>%
  ggplot(aes(x = nCount_RNA)) +
  geom_density(alpha = 0.2) +
  scale_x_log10() +
  theme_classic() +
  ylab("Cell Density")
```

We remove doublets in a droplet-based experiments.

```{r}
if (protocol[2] == "droplet") {

  rds <- file.path(cache_dir, "doublets.rds")
  if (file.exists(rds)) {
    doublets <- readRDS(rds)
  } else {
    sce <- SingleCellExperiment(list(counts = GetAssayData(seurat)))
    sce <- scDblFinder(sce, BPPARAM = MulticoreParam(availableCores()))
    doublets <- sce$scDblFinder.class
    rm(sce)
    saveRDS(doublets, file = rds)
  }

  seurat <- seurat[ , doublets == "singlet"]
  seurat

  print(
    VlnPlot(
      seurat,
      features = c("nFeature_RNA", "nCount_RNA", "percent_mt", "percent_ribo"),
      ncol = 4
    )
  )

  print(
    seurat@meta.data %>%
      ggplot(aes(x = nFeature_RNA)) +
      geom_density(alpha = 0.2) +
      scale_x_log10() +
      theme_classic() +
      ylab("Cell Density")
  )

  print(
    seurat@meta.data %>%
      ggplot(aes(x = nCount_RNA)) +
      geom_density(alpha = 0.2) +
      scale_x_log10() +
      theme_classic() +
      ylab("Cell Density")
  )
}
```

# Cluster Pipeline

We run a pipeline that applies normalization and scaling, dimensionality reduction, and clustering.

```{r}
sub_name <- "prep"
seurat <- cluster_pipeline(
  seurat, cache_dir = cache_dir,
  sub_name = sub_name, protocol = protocol,
  vars_to_regress = vars_to_regress, parallel_override = parallel_override,
  cc = FALSE, res_divider = 50000
)
seurat

for (i in seq_along(metadata_to_plot)) {
  if (
    "no_legend" %in% metadata_to_plot[[i]] &&
      "no_label" %in% metadata_to_plot[[i]]
  ) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]
      ) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2",
        color = names(metadata_to_plot)[i], type = "cat"
      ) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]
      ) +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2",
        color = names(metadata_to_plot)[i], type = "cat"
      ) +
        ggtitle(names(metadata_to_plot)[i])
    )
  }
}

if (protocol[4] == "umis") {
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nFeature_SCT", type = "cont"
    ) + ggtitle("nFeature_SCT")
  )
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nCount_SCT", type = "cont"
    ) + ggtitle("nCount_SCT")
  )
} else if (protocol[4] == "reads") {
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nFeature_RNA", type = "cont"
    ) + ggtitle("nFeature_RNA")
  )
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nCount_RNA", type = "cont"
    ) + ggtitle("nCount_RNA")
  )
}

red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "percent_mt", type = "cont"
) + ggtitle("percent_mt")
red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "percent_ribo", type = "cont"
) + ggtitle("percent_ribo")
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
